seahorn/jobs/linked_list_back/aws_linked_list_back_harness.c

```diff
@@ -1,30 +1,41 @@
 
+
 #include <seahorn/seahorn.h>
 #include <aws/common/linked_list.h>
 #include <linked_list_helper.h>
 
 
 
 int main() {
-    /* data structure */
-    struct aws_linked_list list;
-    size_t length;
-    ensure_linked_list_is_allocated(&list, MAX_LINKED_LIST_ITEM_ALLOCATION, &length);
-
-    /* Keep the old last node of the linked list */
-    struct aws_linked_list_node *old_last = list.tail.prev;
-
-    /* Assume the preconditions. The function requires that list != NULL */
-    assume(!aws_linked_list_empty(&list));
-
-    /* perform operation under verification */
-    struct aws_linked_list_node *back = aws_linked_list_back(&list);
-
-    /* assertions */
-    sassert(sea_aws_linked_list_is_valid(&list, length));
-    sassert(aws_linked_list_node_prev_is_valid(back));
-    sassert(aws_linked_list_node_next_is_valid(back));
-    sassert(back == old_last);
-
-    return 0;
+  /* data structure */
+  struct aws_linked_list list;
+  struct aws_linked_list_node to_add;
+  // struct aws_linked_list_node node1;
+  // struct aws_linked_list_node *nnode;
+  size_t size;
+
+  sea_nd_init_linked_list_from_back(&list, &size);
+  assume(size > 0);
+
+  // TODO: add seahorn support that 1) struct aws_linked_list is never written
+  // to 2) nnode is never written to
+  struct aws_linked_list_node *old_head = &list.head;
+  struct aws_linked_list_node *old_tail = &list.tail;
+  struct aws_linked_list_node *pnode_old = old_tail->prev;
+  struct aws_linked_list_node *pnode_old_next = pnode_old->next;
+  struct aws_linked_list_node *pnode_old_prev = pnode_old->prev;
+
+  /* perform operation under verification */
+  struct aws_linked_list_node *back = aws_linked_list_back(&list);
+
+  /* assertions */
+  sassert(&list.head == old_head);
+  sassert(&list.tail == old_tail);
+  sassert(list.tail.prev == pnode_old);
+  sassert(list.head.next == old_head->next);
```

Comment:
this is wrong. You are checking that the current value of `list.head.next` is equal to itself!
You need to store the value of `list.head.next` before the function call. Now you are storing only the pointer to `&list`. Since assertion above already states that `&list.head = old_head` you know that the list has not changed.

Also, assertions of the form `&list.head == old_head` are meaningless because the function under test can neither change the address of the local variable `list`, nor can it change the value of a local variable `old_head`. So the assertions are true vacuously -- i.e., independently of the code in line 29.

I will stop reviewing harnesses now assuming that the same problem appears all over the place.

We need to create functions to capture footprint and compare since otherwise a bug will appear in many places. Please suggest a design before implementing it

Reply:
I'm thinking the following:

We call the following helper functions are a linked list is created and save state of `node1` and `node2`. We then work only with the nodes these helper functions have returned - to eliminate bugs like the above.
```
// HEAD ... node 2 <-> node 1 <-> TAIL
// node 1 is always non-null, node 2 is NULL if size < 2
// node 1 is HEAD, if size = 0
save_state_from_tail(list, size, *node1, *node2)

// HEAD <-> node 1 <-> node2 ... TAIL
// node 1 is always non-null, node 2 is NULL if size < 2
// node 1 is TAIL, if size = 0
save_state_from_head(list, size, *node1, *node2)
```

Reply:
I don't see (yet) why do we need to mix creation and saving state functions. Unless this is needed for performance, I prefer to have them separate.

Reply:
I think I wasn't clear. We will have two functions

sea_nd_init_linked_list_*
save_state_from_*